### PR TITLE
Refactor `parser` and add comprehensive edge case tests for it

### PR DESCRIPTION
**Pull request by a-hilaly**
This patch refactors the `parser.Parse*` functions, by split the validation
logic from the extraction logic. Now we use helper functions for schema validation
and type checking. Allowing reading to distinguish recursive walk logic, from
other mechanics used for parsing.

- Refactor `parseResource` function to use type specific parsing functions
- Introduce `TestParserEdgeCases` to cover various type mimatches and
  error conditions..
- Add `TestParseWithExpectedSchema` to validate schema extraction from
  objects

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: internal/typesystem/parser/parser.go [hunk 1]
```diff
@@ -57,129 +57,150 @@ func ParseResource(resource map[string]interface{}, resourceSchema *spec.Schema)
 // from a resource. It uses a depthh first search to traverse the resource and
 // extract expressions from string fields
 func parseResource(resource interface{}, schema *spec.Schema, path string) ([]ExpressionField, error) {
```
Comment (a-hilaly) [inline internal/typesystem/parser/parser.go:59]: i'm still mixed on whether we need to sort the expression fields by path. The order of expressions isn't guaranteed here.

Comment (a-hilaly) [reply internal/typesystem/parser/parser.go:59]: The graph runtime doesn't really care about the order of expressions. Probably something to deal with later if it becomes problematic.
